### PR TITLE
Simplify `mithril-client` CLI

### LINKED ISSUE #1406: Simplify Mithril client CLI download command
## Why
Some part of the code of the `mithril-client-cli` and `mithril-client` are redundant or difficult to read. We want to simplify them.

## What 
We want to refactor the `mithril-client` and `mithril-client-cli`:
- Move `add_statistic` responsibility from the cli to the library
- Make `snapshot download` command function more readable by splitting it

## How

- [x] Move `add_statistic` to `mithril-client`
- [x] Modify example and guide for snapshot download with the new `add_statistic` function
- [x] Remove the unused duplicated code in `mithril-client`
- [x] Refactor `snapshot download` command function with a split by reported steps
- [ ] Run [Mithril Client multi-platform test](https://github.com/input-output-hk/mithril/actions/workflows/test-client.yml)

## Later
- Inform users of the current `mithril-client` that `add_statistic` can be called

### PR DESCRIPTION
**Pull request by dlachaumepalo**
## Content
This PR includes an update of `mithril-client` CLI and library.

- `add_statistics` function has been moved from `mithril-client` CLI to the library
- Library examples and documentations have been updated (Rust documentation, user guide)
- Integration tests of the library have been updated and enhanced
- `snapshot download` command has been refactor (split in functions by reported step)

## Pre-submit checklist

- Branch
  - [x] Tests are provided (if possible)
  - [x] Crates versions are updated (if relevant)
  - [x] Commit sequence broadly makes sense
  - [x] Key commits have useful messages
- PR
  - [x] No clippy warnings in the CI
  - [x] Self-reviewed the diff
  - [x] Useful pull request description
  - [x] Reviewer requested
- Documentation
  - [x] Update README file (if relevant)
  - [x] Update documentation website (if relevant)
  - [ ] Add dev blog post (if relevant)

## Comments
Regarding @jpraynaud [comment](https://github.com/input-output-hk/mithril/pull/1413#pullrequestreview-1790895435), we keep the tests as they are for now. If necessary, we'll improve them later as they don't have a negative impact on the project, apart from making the code in tests more cumbersome.

## Issue(s)
Closes #1406

### CODE REVIEW

File: mithril-client/tests/extensions/fake.rs [hunk 1]
```diff
@@ -21,10 +26,41 @@ impl FakeCertificateVerifier {
...
+
+        calls.clone()
+    }
+
+    pub async fn get_last_call(&self) -> Option<String> {
+        self.get_calls().await.last().cloned()
+    }
+
+    pub async fn store_call_and_return_value(
+        _params: Vec<String>,
```
Comment (jpraynaud) [inline mithril-client/tests/extensions/fake.rs:51]: Can be removed if it is not used?

**Review by jpraynaud — CHANGES_REQUESTED** [review]
It looks like we are testing twice the same thing with the `FakeAggregator` modifications: 
- A call to the wrong route would not return the expected values (e.g. we would receive `Certificate` when we expect `Snapshot`) and thus the tests would fail
- It means that there is no real need to test the url/payload called on the aggregator
- The only route where this is not the case is `/statistics/snapshot` for which we'd like to know if the remote call has been properly executed
- We'd rather verify only the call on the `/statistics/snapshot` is done when the restoration of the snapshot is successful

Also, I left few minor comments below for typos and naming.

**Review by ghubertpalo — APPROVED** [review]

File: mithril-client-cli/src/commands/snapshot/download.rs [hunk 2]
```diff
@@ -92,39 +92,35 @@ impl SnapshotDownloadCommand {
...
-            &progress_printer,
-            &certificate,
-            &db_dir,
-        )
-        .await?;
+        let message =
+            Self::compute_snapshot_message(4, &progress_printer, &certificate, &db_dir).await?;
 
-        SnapshotDownloadCommand::verifying_snapshot_signature(
+        Self::verifying_snapshot_signature(
```
Comment (jpraynaud) [inline mithril-client-cli/src/commands/snapshot/download.rs:123]: You forgot this one :slightly_smiling_face: 
```suggestion
        Self::verify_snapshot_signature(
```

**Review by jpraynaud — APPROVED** [review]
LGTM

One more typo to fix and I reopened two previous comments still to be fixed before merging

File: mithril-client-cli/src/commands/snapshot/download.rs (hunk 2 above)
Comment (dlachaumepalo) [reply mithril-client-cli/src/commands/snapshot/download.rs:123]: Oops I'll fix that right away 🤓, thanks!

File: mithril-client-cli/src/commands/snapshot/download.rs [hunk 3]
```diff
@@ -117,30 +181,35 @@ impl SnapshotDownloadCommand {
...
             .with_context(|| {
                 format!(
                     "Snapshot Service can not download and verify the snapshot for digest: '{}'",
-                    self.digest
+                    snapshot_message.digest
                 )
             })?;
 
         // The snapshot download does not fail if the statistic call fails.
         // It would be nice to implement tests to verify the behavior of `add_statistics`
```
[the remark this answers is not in the document]
Comment (dlachaumepalo) [reply mithril-client-cli/src/commands/snapshot/download.rs:198]: It's already fixed in this commit. Weird, is it a display issue on github caused by squashing and rebase interactive?
https://github.com/input-output-hk/mithril/pull/1413/commits/af2ed8e4caa10236b149296b60f1666b11945498

File: mithril-client-cli/src/commands/snapshot/download.rs [hunk 4]
```diff
@@ -90,24 +92,86 @@ impl SnapshotDownloadCommand {
...
         })?;
 
+        Ok(())
+    }
+
+    async fn fetching_certificate_and_verifying_chain(
+        step_number: u16,
+        progress_printer: &ProgressPrinter,
+        client: &Client,
+        snapshot_message: &Snapshot,
```
[the remark this answers is not in the document]
Comment (dlachaumepalo) [reply mithril-client-cli/src/commands/snapshot/download.rs]: It's already fixed in this commit. Weird, is it a display issue on github caused by squashing and rebase interactive?
https://github.com/input-output-hk/mithril/pull/1413/commits/af2ed8e4caa10236b149296b60f1666b11945498

**Pull request merged**